## Date:25/05/2020

# Capstone Project – The Battle of Neighborhoods | Finding the Common places in Toronto which needs to be avoided due to Covid-19 crisis

### 1. Installing and Importing Python Libraries and Dependencies

In [1]:
!pip install geocoder
!pip install folium

Importing Libraries

In [2]:
import pandas as pd
import requests
import numpy as np
import geocoder
import folium
import requests 
import matplotlib.cm as cm
import matplotlib.colors as colors
import json
import xml
import matplotlib.pyplot as plt
%matplotlib inline
import warnings
warnings.filterwarnings("ignore")

from pandas.io.json import json_normalize 
from sklearn.cluster import KMeans
from geopy.geocoders import Nominatim 
from bs4 import BeautifulSoup

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

print("All Required Libraries Imported!")

All Required Libraries Imported!


### 2.  Data Extraction and Cleaning

Using BeautifulSoup Scraping List of Postal Codes of Given Wikipedia Page. 
Link: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

In [3]:
List_url='https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'
source = requests.get(List_url).text

soup = BeautifulSoup(source, 'xml')
table=soup.find('table')
column_names=['Postalcode','Borough','Neighborhood']
df = pd.DataFrame(columns=column_names)

Converting content of PostalCode HTML table as dataframe

In [4]:
for tr_cell in table.find_all('tr'):
    row_data=[]
    for td_cell in tr_cell.find_all('td'):
        row_data.append(td_cell.text.strip())
    if len(row_data)==3:
        df.loc[len(df)] = row_data
        df.head()

In [5]:
# clean dataframe 
df=df[df['Borough']!='Not assigned']

In [6]:
temp_df=df.groupby('Postalcode')['Neighborhood'].apply(lambda x: "%s" % ', '.join(x))
temp_df=temp_df.reset_index(drop=False)
temp_df.rename(columns={'Neighborhood':'Neighborhood_joined'},inplace=True)

In [7]:
df_merge = pd.merge(df, temp_df, on='Postalcode')

In [8]:
df_merge.drop(['Neighborhood'],axis=1,inplace=True)

In [9]:
df_merge.drop_duplicates(inplace=True)

In [10]:
df_merge.rename(columns={'Neighborhood_joined':'Neighborhood'},inplace=True)

In [11]:
df_merge.head()

Postalcode           Borough                                 Neighborhood
0        M3A        North York                                    Parkwoods
1        M4A        North York                             Victoria Village
2        M5A  Downtown Toronto                    Regent Park, Harbourfront
3        M6A        North York             Lawrence Manor, Lawrence Heights
4        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [12]:
df_merge.shape

(103, 3)

In [13]:
def get_geocode(postal_code):
    # initialize your variable to None
    lat_lng_coords = None
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
        lat_lng_coords = g.latlng
    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    return latitude,longitude

In [14]:
# Retrieving Postal Code Co-ordinates
geo_df=pd.read_csv('http://cocl.us/Geospatial_data')

In [15]:
geo_df.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

In [16]:
# Adding Columns Latitude & Longitude
geo_df.rename(columns={'Postal Code':'Postalcode'},inplace=True)
geo_merged = pd.merge(geo_df, df_merge, on='Postalcode')

In [17]:
geo_merged.head(20)

Postalcode   Latitude  Longitude      Borough  \
0         M1B  43.806686 -79.194353  Scarborough   
1         M1C  43.784535 -79.160497  Scarborough   
2         M1E  43.763573 -79.188711  Scarborough   
3         M1G  43.770992 -79.216917  Scarborough   
4         M1H  43.773136 -79.239476  Scarborough   
5         M1J  43.744734 -79.239476  Scarborough   
6         M1K  43.727929 -79.262029  Scarborough   
7         M1L  43.711112 -79.284577  Scarborough   
8         M1M  43.716316 -79.239476  Scarborough   
9         M1N  43.692657 -79.264848  Scarborough   
10        M1P  43.757410 -79.273304  Scarborough   
11        M1R  43.750072 -79.295849  Scarborough   
12        M1S  43.794200 -79.262029  Scarborough   
13        M1T  43.781638 -79.304302  Scarborough   
14        M1V  43.815252 -79.284577  Scarborough   
15        M1W  43.799525 -79.318389  Scarborough   
16        M1X  43.836125 -79.205636  Scarborough   
17        M2H  43.803762 -79.363452   North York   
18        M2J  43.778517 -79.346556   North York   
19        M2K  43.786947 -79.385975   North York   

                                         Neighborhood  
0                                      Malvern, Rouge  
1              Rouge Hill, Port Union, Highland Creek  
2                   Guildwood, Morningside, West Hill  
3                                              Woburn  
4                                           Cedarbrae  
5                                 Scarborough Village  
6         Kennedy Park, Ionview, East Birchmount Park  
7                     Golden Mile, Clairlea, Oakridge  
8     Cliffside, Cliffcrest, Scarborough Village West  
9                         Birch Cliff, Cliffside West  
10  Dorset Park, Wexford Heights, Scarborough Town...  
11                                  Wexford, Maryvale  
12                                          Agincourt  
13            Clarks Corners, Tam O'Shanter, Sullivan  
14  Milliken, Agincourt North, Steeles East, L'Amo...  
15                      Steeles West, L'Amoreaux West  
16                                        Upper Rouge  
17                                  Hillcrest Village  
18                       Fairview, Henry Farm, Oriole  
19                                    Bayview Village

In [18]:
geo_data=geo_merged[['Postalcode','Borough','Neighborhood','Latitude','Longitude']]
geo_data.head(20)

Postalcode      Borough                                       Neighborhood  \
0         M1B  Scarborough                                     Malvern, Rouge   
1         M1C  Scarborough             Rouge Hill, Port Union, Highland Creek   
2         M1E  Scarborough                  Guildwood, Morningside, West Hill   
3         M1G  Scarborough                                             Woburn   
4         M1H  Scarborough                                          Cedarbrae   
5         M1J  Scarborough                                Scarborough Village   
6         M1K  Scarborough        Kennedy Park, Ionview, East Birchmount Park   
7         M1L  Scarborough                    Golden Mile, Clairlea, Oakridge   
8         M1M  Scarborough    Cliffside, Cliffcrest, Scarborough Village West   
9         M1N  Scarborough                        Birch Cliff, Cliffside West   
10        M1P  Scarborough  Dorset Park, Wexford Heights, Scarborough Town...   
11        M1R  Scarborough                                  Wexford, Maryvale   
12        M1S  Scarborough                                          Agincourt   
13        M1T  Scarborough            Clarks Corners, Tam O'Shanter, Sullivan   
14        M1V  Scarborough  Milliken, Agincourt North, Steeles East, L'Amo...   
15        M1W  Scarborough                      Steeles West, L'Amoreaux West   
16        M1X  Scarborough                                        Upper Rouge   
17        M2H   North York                                  Hillcrest Village   
18        M2J   North York                       Fairview, Henry Farm, Oriole   
19        M2K   North York                                    Bayview Village   

     Latitude  Longitude  
0   43.806686 -79.194353  
1   43.784535 -79.160497  
2   43.763573 -79.188711  
3   43.770992 -79.216917  
4   43.773136 -79.239476  
5   43.744734 -79.239476  
6   43.727929 -79.262029  
7   43.711112 -79.284577  
8   43.716316 -79.239476  
9   43.692657 -79.264848  
10  43.757410 -79.273304  
11  43.750072 -79.295849  
12  43.794200 -79.262029  
13  43.781638 -79.304302  
14  43.815252 -79.284577  
15  43.799525 -79.318389  
16  43.836125 -79.205636  
17  43.803762 -79.363452  
18  43.778517 -79.346556  
19  43.786947 -79.385975

In [19]:
# Getting the Toronto data
toronto_data=geo_data[geo_data['Borough'].str.contains("Toronto")]
toronto_data.head()

Postalcode          Borough                    Neighborhood   Latitude  \
37        M4E     East Toronto                     The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
43        M4M     East Toronto                 Studio District  43.659526   
44        M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  
37 -79.293031  
41 -79.352188  
42 -79.315572  
43 -79.340923  
44 -79.388790

### 3. Venues/Locations

In [20]:
CLIENT_ID = 'F12555BXBRL5V1PI3OAK24A0TW0H1JY43PJ0QN35KB2CXZVN'
CLIENT_SECRET = 'IWXLZ04MFGYRIGU4XD1LJ5DYLPLDGTK3SDXRSR0W4G3G5HDW' 
VERSION = '20200511'

In [21]:
def getNearbyVenues(names, latitudes, longitudes):
    radius=500
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [22]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West,  Lawrence Park
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West, Forest Hill Road Park
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Stn A PO Boxes
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High

In [23]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  \
0                   The Beaches              43.676357   
1                   The Beaches              43.676357   
2                   The Beaches              43.676357   
3                   The Beaches              43.676357   
4  The Danforth West, Riverdale              43.679557   

   Neighborhood Longitude                               Venue  Venue Latitude  \
0              -79.293031                   Glen Manor Ravine       43.676821   
1              -79.293031  The Big Carrot Natural Food Market       43.678879   
2              -79.293031                 Grover Pub and Grub       43.679181   
3              -79.293031                       Upper Beaches       43.680563   
4              -79.352188                       MenEssentials       43.677820   

   Venue Longitude     Venue Category  
0       -79.293942              Trail  
1       -79.297734  Health Food Store  
2       -79.297215                Pub  
3       -79.292869       Neighborhood  
4       -79.351265     Cosmetics Shop

In [24]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            56   
Brockton, Parkdale Village, Exhibition Place                           23   
Business reply mail Processing Centre, South Ce...                     17   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     62   
Christie                                                               17   
Church and Wellesley                                                   76   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                        7   
Dufferin, Dovercourt Village                                           14   
First Canadian Place, Underground city                                100   
Forest Hill North & West, Forest Hill Road Park                         5   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          23   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              58   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               43   
Moore Park, Summerhill East                                             3   
North Toronto West,  Lawrence Park                                     23   
Parkdale, Roncesvalles                                                 14   
Queen's Park, Ontario Provincial Government                            34   
Regent Park, Harbourfront                                              48   
Richmond, Adelaide, King                                               92   
Rosedale                                                                4   
Roselawn                                                                2   
Runnymede, Swansea                                                     37   
St. James Town                                                         77   
St. James Town, Cabbagetown                                            46   
Stn A PO Boxes                                                         92   
Studio District                                                        40   
Summerhill West, Rathnelly, South Hill, Forest ...                     16   
The Annex, North Midtown, Yorkville                                    23   
The Beaches                                                             4   
The Danforth West, Riverdale                                           42   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         34   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             56   
Brockton, Parkdale Village, Exhibition Place                            23   
Business reply mail Processing Centre, South Ce...                      17   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      62   
Christie                                                                17   
Church and Wellesley                                                    76   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                    

In [25]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.drop(['Neighborhood'],axis=1,inplace=True) 
toronto_onehot.insert(loc=0, column='Neighborhood', value=toronto_venues['Neighborhood'] )
toronto_onehot.shape

(1605, 235)

In [39]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Afghan Restaurant  \
0                                        Berczy Park                0.0   
1       Brockton, Parkdale Village, Exhibition Place                0.0   
2  Business reply mail Processing Centre, South C...                0.0   
3  CN Tower, King and Spadina, Railway Lands, Har...                0.0   
4                                 Central Bay Street                0.0   

    Airport  Airport Food Court  Airport Gate  Airport Lounge  \
0  0.000000            0.000000      0.000000        0.000000   
1  0.000000            0.000000      0.000000        0.000000   
2  0.000000            0.000000      0.000000        0.000000   
3  0.066667            0.066667      0.066667        0.133333   
4  0.000000            0.000000      0.000000        0.000000   

   Airport Service  Airport Terminal  American Restaurant  Antique Shop  \
0         0.000000          0.000000                  0.0           0.0   
1         0.000000          0.000000                  0.0           0.0   
2         0.000000          0.000000                  0.0           0.0   
3         0.133333          0.133333                  0.0           0.0   
4         0.000000          0.000000                  0.0           0.0   

   Aquarium  Art Gallery  Art Museum  Arts & Crafts Store  Asian Restaurant  \
0       0.0     0.017857    0.000000                  0.0               0.0   
1       0.0     0.000000    0.000000                  0.0               0.0   
2       0.0     0.000000    0.000000                  0.0               0.0   
3       0.0     0.000000    0.000000                  0.0               0.0   
4       0.0     0.000000    0.016129                  0.0               0.0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                 0.0       0.000000   0.017857         0.0    0.017857   
1                 0.0       0.000000   0.000000         0.0    0.000000   
2                 0.0       0.058824   0.000000         0.0    0.000000   
3                 0.0       0.000000   0.000000         0.0    0.000000   
4                 0.0       0.000000   0.000000         0.0    0.000000   

     Bakery  Bank       Bar  Baseball Stadium  Basketball Stadium     Beach  \
0  0.035714   0.0  0.000000               0.0            0.017857  0.017857   
1  0.043478   0.0  0.043478               0.0            0.000000  0.000000   
2  0.000000   0.0  0.000000               0.0            0.000000  0.000000   
3  0.000000   0.0  0.066667               0.0            0.000000  0.000000   
4  0.000000   0.0  0.032258               0.0            0.000000  0.000000   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant    Bistro  \
0              0.0  0.035714         0.0                 0.0  0.017857   
1              0.0  0.000000         0.0                 0.0  0.000000   
2              0.0  0.000000         0.0                 0.0  0.000000   
3              0.0  0.000000         0.0                 0.0  0.000000   
4              0.0  0.000000         0.0                 0.0  0.000000   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0       0.000000        0.0  0.000000                   0.0        0.017857   
1       0.000000        0.0  0.000000                   0.0        0.086957   
2       0.000000        0.0  0.000000                   0.0        0.000000   
3       0.066667        0.0  0.066667                   0.0        0.000000   
4       0.000000        0.0  0.000000                   0.0        0.000000   

    Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0  0.000000         0.000000       0.0      0.000000       0.000000       0.0   
1  0.000000         0.000000       0.0      0.000000       0.043478       0.0   
2  0.058824         0.000000       0.0      0.000000       0.058824       0.0   
3  0.000000         0.000000       0.0      0.000000       0.000000       0.0   
4  0.000000         0

### 4. Common venues near neighborhood

In [40]:
# Getting top venues
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [41]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  Business reply mail Processing Centre, South C...           Yoga Studio   
3  CN Tower, King and Spadina, Railway Lands, Har...        Airport Lounge   
4                                 Central Bay Street           Coffee Shop   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0          Cocktail Bar                  Café              Beer Bar   
1        Breakfast Spot             Nightclub           Coffee Shop   
2           Pizza Place            Smoke Shop               Brewery   
3       Airport Service      Airport Terminal       Harbor / Marina   
4    Italian Restaurant        Sandwich Place                  Café   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                Bakery    Seafood Restaurant           Cheese Shop   
1          Climbing Gym         Burrito Place    Italian Restaurant   
2                   Spa         Burrito Place        Farmers Market   
3      Sculpture Garden           Coffee Shop         Boat or Ferry   
4          Burger Joint        Ice Cream Shop   Japanese Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0            Restaurant                   Pub         Farmers Market  
1          Intersection               Stadium             Restaurant  
2  Fast Food Restaurant            Restaurant     Light Rail Station  
3              Boutique                   Bar           Airport Gate  
4       Thai Restaurant                   Bar        Bubble Tea Shop

### 5. K-Means Clustering Approach


In [42]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [43]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() 

Postalcode          Borough                    Neighborhood   Latitude  \
37        M4E     East Toronto                     The Beaches  43.676357   
41        M4K     East Toronto    The Danforth West, Riverdale  43.679557   
42        M4L     East Toronto  India Bazaar, The Beaches West  43.668999   
43        M4M     East Toronto                 Studio District  43.659526   
44        M4N  Central Toronto                   Lawrence Park  43.728020   

    Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
37 -79.293031               0                 Trail                   Pub   
41 -79.352188               0      Greek Restaurant           Coffee Shop   
42 -79.315572               0  Fast Food Restaurant            Steakhouse   
43 -79.340923               0                  Café           Coffee Shop   
44 -79.388790               3                  Park    Dim Sum Restaurant   

   3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
37     Health Food Store   Distribution Center        Department Store   
41    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
42         Burrito Place                   Pub              Restaurant   
43             Gastropub                Bakery                 Brewery   
44              Bus Line           Swim School             Event Space   

   6th Most Common Venue        7th Most Common Venue  \
37          Dessert Shop           Dim Sum Restaurant   
41            Restaurant              Bubble Tea Shop   
42          Intersection            Fish & Chips Shop   
43   American Restaurant                  Yoga Studio   
44     Electronics Store  Eastern European Restaurant   

      8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
37                    Diner        Discount Store            Yoga Studio  
41                   Bakery                   Pub            Pizza Place  
42       Italian Restaurant           Pizza Place                    Gym  
43  Comfort Food Restaurant    Seafood Restaurant         Sandwich Place  
44               Donut Shop      Doner Restaurant                Dog Run

In [44]:
neighborhoods_venues_sorted.head()

Cluster Labels                                       Neighborhood  \
0               0                                        Berczy Park   
1               0       Brockton, Parkdale Village, Exhibition Place   
2               0  Business reply mail Processing Centre, South C...   
3               0  CN Tower, King and Spadina, Railway Lands, Har...   
4               0                                 Central Bay Street   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop          Cocktail Bar                  Café   
1                  Café        Breakfast Spot             Nightclub   
2           Yoga Studio           Pizza Place            Smoke Shop   
3        Airport Lounge       Airport Service      Airport Terminal   
4           Coffee Shop    Italian Restaurant        Sandwich Place   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0              Beer Bar                Bakery    Seafood Restaurant   
1           Coffee Shop          Climbing Gym         Burrito Place   
2               Brewery                   Spa         Burrito Place   
3       Harbor / Marina      Sculpture Garden           Coffee Shop   
4                  Café          Burger Joint        Ice Cream Shop   

  7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
0           Cheese Shop            Restaurant                   Pub   
1    Italian Restaurant          Intersection               Stadium   
2        Farmers Market  Fast Food Restaurant            Restaurant   
3         Boat or Ferry              Boutique                   Bar   
4   Japanese Restaurant       Thai Restaurant                   Bar   

  10th Most Common Venue  
0         Farmers Market  
1             Restaurant  
2     Light Rail Station  
3           Airport Gate  
4        Bubble Tea Shop

In [45]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6534817, -79.3839347.


### Map of Toronto

In [47]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

Conclusion: In this project, using k-means cluster algorithm I separated the neighborhood into 10(Ten) different clusters and for 103 different lattitude and logitude from dataset, which have common venues. Based on Top 10 venues these places can be avoided during Covid-19 crisis.